### ISSUE TITLE
Logic to limit block and functionality based on user capabilities/Jetpack connection

### ISSUE THREAD

**Issue by SiobhyB** [wordpress-mobile/gutenberg-mobile#5475]
**Description of the problem.**

The VideoPress block should only be available to users who have the necessary capabilities to use the block. Users need to have an active Jetpack connection and also have a plan that allows for VideoPress uploads.

**Proposed solution.**

On the web, the following function is used for this purpose:

https://github.com/Automattic/jetpack/blob/8fa35bced2096201a26744703a00bb31fde7fc49/projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php#L62-L97

We should implement something similar for Gutenberg Mobile.

**Comment by SiobhyB** [comment]
Notes so far. 

### Existing web logic

On the web's side, the `check_videopress_availability` function exists [here](https://github.com/Automattic/jetpack/blob/8fa35bced2096201a26744703a00bb31fde7fc49/projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php#L62-L97) in the `class.videopress-gutenberg.php` file. It sets the block's availability based on the following checks:

- If the user's plan doesn't support VideoPress, then availability is set to `false` straightaway. 
- If the user's site is not hosted on WordPress.com, the Jetpack connection status is checked.
   - If there isn't an active connection, availability is set to `false`.
   - If there is an active connection, there is a further check to see if the plugin's VideoPress module is enabled and availability is only set to `true` if there is.
- If the user's site is hosted on WordPress.com and they have a plan that supports VideoPress, availability is set to `true`.

### Existing native logic

From the native side, we manually set each Jetpack block's availability [here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/5b6765f59fc0c795a89b14c527014223f9275824/src/jetpack-editor-setup.js#L38-L56) in Gutenberg Mobile's `jetpack-editor-setup.js` file. There are no server side calls to limit block availability in the React Native codebase. Instead, there are currently checks in place [here](https://github.com/wordpress-mobile/WordPress-Android/blob/261be1f09e784b03fddc9e38d8ebe001a57bb993/WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java#L351-L353) on Android and [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/9f6320d85ecefa574bb32b85e976c513fd962dd8/WordPress/Classes/Models/Blog.m#L718-L721) on iOS to limit the block to simple WordPress.com sites.

### Next steps for native logic

After spending some time experimenting, I'd like to propose that we iterate on the existing logic in the Android and iOS codebases as follows for each of the project's stages:
   - MVP: Detect whether the user's plan supports VideoPress for simple WordPress.com sites.
   - Advanced: Add support for Atomic sites and check for active Jetpack connection (the calls needed to check for an active connection already exist in both the Android and iOS codebases).

I was originally planning to find a way to implement this logic within the React Native codebase, as it didn't feel optimal that the availability was being manually set there. After taking a step back to consider this, though, using the existing logic in Android and iOS feels like it may be the better option. The calls we need are already being made in those codebases and I don't know if there's a good reason to stray from that.

@fluiddot, @jhnstn, I wanted to double-check with both of you. Am I missing anything here? Do you have any concerns about my proposed path forward? Thanks in advance for the confidence check!

**Comment by fluiddot** [comment]
> I was originally planning to find a way to implement this logic within the React Native codebase, as it didn't feel optimal that the availability was being manually set there. After taking a step back to consider this, though, using the existing logic in Android and iOS feels like it may be the better option. The calls we need are already being made in those codebases and I don't know if there's a good reason to stray from that.

@SiobhyB As you mentioned, the block's availability for the web version is done in the backend, in the Gutenberg Mobile case, as we don't communicate directly with the backend, we could consider the native side/host app as our backend. Hence, the option of updating the existing logic on the native side seems the better option.

> MVP: Detect whether the user's plan supports VideoPress for simple WordPress.com sites.

I'm wondering if there's any case where VideoPress shouldn't be available on simple WPCOM sites. As far as I checked, VideoPress is available for all plans including the free plan. In the web version, what's not available in the free plan is uploading videos. However, this is not the case in the app, as we allow uploads but with some limitations (pbArwn-3mr-p2#comment-4556).

> Advanced: Add support for Atomic sites and check for active Jetpack connection (the calls needed to check for an active connection already exist in both the Android and iOS codebases).

For this case, I understand we'd need to check the Jetpack connection. I haven't tried to disconnect Jetpack on an Atomic site but I guess it's possible 🤔.

**Comment by SiobhyB** [comment]
Thanks Carlos for the confidence check!

> I'm wondering if there's any case where VideoPress shouldn't be available on simple WPCOM sites. 

I was leaning towards matching the web's behaviour of the block only being _usable_ for free and Personal plans with this task. Thinking through your points, however, I agree that the fact free video uploads are a long-standing issue in the apps makes this less of a priority. It'd probably make more sense to address this when we begin considering an upgrade path. Does that match what you're thinking?

I've added a new `Future` column in the project board with a couple of tasks to keep track of things that we aren't going to address right now, but will in the future:

<img width="537" alt="Screenshot 2023-03-02 at 10 49 36" src="https://user-images.githubusercontent.com/2998162/222407670-3e9cafd9-5cda-419d-81b5-b563e585c4db.png">

> For this case, I understand we'd need to check the Jetpack connection. I haven't tried to disconnect Jetpack on an Atomic site but I guess it's possible 🤔.

The connection only gets broken when something goes wrong on Atomic sites, usually due to a faulty plugin. It's not possible for users to toggle the connection off (HEs have to fix broken connections on our side). 

So, I think for the second stage of the project, we can use the existing `isConnected` Jetpack checks on iOS and Android to open up the block to more users.

With all of the above considered, it doesn't feel like there's anything more to do for this task at this current MVP stage. The block is already limited to all simple sites. Do you agree? If so, I'll update the project thread.

**Comment by fluiddot** [comment]
> I was leaning towards matching the web's behaviour of the block only being usable for free and Personal plans with this task. Thinking through your points, however, I agree that the fact free video uploads are a long-standing issue in the apps makes this less of a priority. It'd probably make more sense to address this when we begin considering an upgrade path. Does that match what you're thinking?

Yep, I agree. My understanding about free video uploads is that this is an important feature we want to deliver in the app, even though it differs from the web behavior. Something I noticed when testing on a free plan is that we don't display any information about this limitation anywhere, even when the upload fails. This is definitely an issue it would be great to fix but not strictly related to the VideoPress project. Nevertheless, I'll add a task as a follow-up.

> I've added a new Future column in the project board with a couple of tasks to keep track of things that we aren't going to address right now, but will in the future:

Great, thanks 🙇 !

> The connection only gets broken when something goes wrong on Atomic sites, usually due to a faulty plugin. It's not possible for users to toggle the connection off (HEs have to fix broken connections on our side).

That's interesting, thanks for the info.

> So, I think for the second stage of the project, we can use the existing isConnected Jetpack checks on iOS and Android to open up the block to more users.

Yeah, we could expand that functionality in the main apps.

> With all of the above considered, it doesn't feel like there's anything more to do for this task at this current MVP stage. The block is already limited to all simple sites. Do you agree? If so, I'll update the project thread.

Yes, I agree. At least for the MVP seems we don't have to do anything else in this task. Thanks @SiobhyB for working in this and sharing your investigations 🙇 !

**Comment by SiobhyB** [comment]
Sounds great, thanks again @fluiddot 🙌 

I'll go ahead to un-assign myself from this task for now, as there issue any further attention needed until the next phase.
